### PR TITLE
Add SPPM render controls

### PR DESCRIPTION
**Pull request by Mango-3**
- Added all render controls for SPPM including max. ray intensity
- Added `Photon Target` object attribute specifier

### CODE REVIEW

**Review by est77 — APPROVED** [review]

File: scripts/appleseedMaya/renderGlobals.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 #
 # This software is released under the MIT license.
 #
-# Copyright (c) 2016-2018 Esteban Tovagliari, The appleseedhq Organization
+# Copyright (c) 2016-2019 Esteban Tovagliari, The appleseedhq Organization
```
Comment (dictoon) [inline scripts/appleseedMaya/renderGlobals.py:8]: Yep, I need to do a pass on all files to update the copyright notices...

File: scripts/appleseedMaya/renderGlobals.py [hunk 2]
```diff
@@ -431,48 +440,127 @@ def create(self):
...
+                        self._addControl(
+                            ui=pm.attrEnumOptionMenuGrp(
+                                label="Direct Lighting",
+                                enumeratedItem=self._getAttributeMenuItems("SPPMLightingMode")),
+                            attrName="SPPMLightingMode")
+                        self._addControl(
+                            ui=pm.checkBoxGrp(label="Caustics"),
+                            attrName="SPPMCaustics")
+                        self._addControl(
+                            ui=pm.checkBoxGrp(label="Image Based Lighting"),
```
Comment (dictoon) [inline scripts/appleseedMaya/renderGlobals.py:502]: Add dash: *Image-Based Lighting*

File: scripts/appleseedMaya/renderGlobals.py [hunk 3]
```diff
@@ -431,48 +440,127 @@ def create(self):
...
+                        limitPhotonTracingBounces = mc.getAttr(
+                                "appleseedRenderGlobals.limitPhotonTracingBounces")
+                        self._addControl(
+                            ui=pm.intFieldGrp(
+                                label="Max Bounces",  columnWidth=(3, 120), columnAlign = (3, 'left'), 
+                                extraLabel ='Photon Tracing', numberOfFields=1, enable=limitPhotonTracingBounces),
+                            attrName="photonTracingBounces")
+                        self._addControl(
+                            ui=pm.intFieldGrp(
+                                label="RR Start Bounce", numberOfFields=1),
```
Comment (dictoon) [inline scripts/appleseedMaya/renderGlobals.py:517]: If space allows, I would suggest omitting abbreviations: *Russian Roulette Start Bounce*

File: scripts/appleseedMaya/renderGlobals.py [hunk 4]
```diff
@@ -431,48 +440,127 @@ def create(self):
...
+                            attrName="SPPMLightingMode")
+                        self._addControl(
+                            ui=pm.checkBoxGrp(label="Caustics"),
+                            attrName="SPPMCaustics")
+                        self._addControl(
+                            ui=pm.checkBoxGrp(label="Image Based Lighting"),
+                            attrName="SPPMEnableIBL")
+                        self._addControl(
+                            ui=pm.checkBoxGrp(
+                                label="Limit PT Bounces", changeCommand=self.__limitPhotonTracingBouncesChanged),
```
Comment (dictoon) [inline scripts/appleseedMaya/renderGlobals.py:506]: If space allows, I would suggest omitting abbreviations: *Limit Photon Tracing Bounces* (I suppose that's what PT means here?)

File: scripts/appleseedMaya/renderGlobals.py [hunk 5]
```diff
@@ -431,48 +440,127 @@ def create(self):
...
+                            ui=pm.intFieldGrp(
+                                label="Light Photons", numberOfFields=1),
+                            attrName="photonTracingLightPhotons")
+                        self._addControl(
+                            ui=pm.intFieldGrp(
+                                label="Environment Photons", numberOfFields=1),
+                            attrName="photonTracingEnvPhotons")
+                        self._addControl(
+                            ui=pm.checkBoxGrp(
+                                label="Limit RE Bounces", changeCommand=self.__limitRadianceEstimationBouncesChanged),
```
Comment (dictoon) [inline scripts/appleseedMaya/renderGlobals.py:529]: *Limit Radiance Estimation Bounces*

File: src/appleseedmaya/renderglobalsnode.cpp [hunk 6]
```diff
@@ -317,19 +340,109 @@ MStatus RenderGlobalsNode::initialize()
...
+    numAttrFn.setMin(0.001);
+    numAttrFn.setMax(100.0);
+    CHECKED_ADD_ATTRIBUTE(m_sppm_radiance_estimation_initial_radius, "radianceEstimationInitialRadius")
+
+    // Radiance estimation max photons.
+    m_sppm_radiance_estimation_max_photons = numAttrFn.create("radianceEstimationMaxPhotons", "radianceEstimationMaxPhotons", MFnNumericData::kInt, 100, &status);
+    numAttrFn.setMin(8);
+    CHECKED_ADD_ATTRIBUTE(m_sppm_radiance_estimation_max_photons, "radianceEstimationMaxPhotons")
+
+    // Radiance estimation search radius reduction factor.
```
Comment (dictoon) [inline src/appleseedmaya/renderglobalsnode.cpp:430]: Yes! That's how we should call Alpha: *Search Radius Reduction Factor*

File: src/appleseedmaya/renderglobalsnode.cpp [hunk 7]
```diff
@@ -346,11 +459,11 @@ MStatus RenderGlobalsNode::initialize()
     numAttrFn.setMin(1);
     CHECKED_ADD_ATTRIBUTE(m_mbDeformSamples, "deformSamples")
 
-    // Shutter Open.
+    // Shutter open.
```
Comment (dictoon) [inline src/appleseedmaya/renderglobalsnode.cpp:462]: I know that's completely unrelated to this PR, but probably worth mentioning anyway: appleseed now supports four times:

  - shutter open begin
  - shutter open end
  - shutter close begin
  - shutter close end

as well as various opening/closing curves for the [begin, end] intervals.

File: src/appleseedmaya/renderglobalsnode.cpp [hunk 8]
```diff
@@ -392,33 +505,32 @@ MStatus RenderGlobalsNode::initialize()
     m_randomPixelOrder = numAttrFn.create("randomPixelOrder", "randomPixelOrder", MFnNumericData::kBoolean, true, &status);
     CHECKED_ADD_ATTRIBUTE(m_randomPixelOrder, "randomPixelOrder")
 
-    // Prefilter Spikes.
+    // Prefilter spikes.
     m_prefilterSpikes = numAttrFn.create("prefilterSpikes", "prefilterSpikes", MFnNumericData::kBoolean, true, &status);
     CHECKED_ADD_ATTRIBUTE(m_prefilterSpikes, "prefilterSpikes")
 
-    // Spike Thereshold.
+    // Spike thereshold.
```
Comment (dictoon) [inline src/appleseedmaya/renderglobalsnode.cpp:512]: *threshold* (the typo was there before)

File: src/appleseedmaya/renderglobalsnode.cpp [hunk 9]
```diff
@@ -593,11 +708,15 @@ void RenderGlobalsNode::applyGlobalsToProject(
...
     }
 
+    // Lighting params.
     int lightingEngine;
     if (AttributeUtils::get(MPlug(globals, m_lightingEngine), lightingEngine))
     {
         if (lightingEngine == 0)
             finalParams.insert_path("lighting_engine", "pt");
+
+        if (lightingEngine == 1)
```
Comment (dictoon) [inline src/appleseedmaya/renderglobalsnode.cpp:718]: Even better: use a `switch`.

File: src/appleseedmaya/renderglobalsnode.cpp [hunk 10]
```diff
@@ -648,6 +767,97 @@ void RenderGlobalsNode::applyGlobalsToProject(
...
         INSERT_PATH_IN_CONFIGS("pt.ibl_env_samples", envSamples)
 
+    // Stochastic progressive photon mapping params.
+    int sppmPhotonType;
+    if (AttributeUtils::get(MPlug(globals, m_sppm_photon_type), sppmPhotonType))
+    {
+        if (sppmPhotonType == 0)
+            INSERT_PATH_IN_CONFIGS("sppm.photon_type", "poly");
+
+        if (sppmPhotonType == 1)
```
Comment (dictoon) [inline src/appleseedmaya/renderglobalsnode.cpp:777]: Use a `switch`.

File: src/appleseedmaya/renderglobalsnode.cpp [hunk 11]
```diff
@@ -648,6 +767,97 @@ void RenderGlobalsNode::applyGlobalsToProject(
...
+            INSERT_PATH_IN_CONFIGS("sppm.photon_type", "mono");
+    }
+
+    int sppmDirectLightingMode;
+    if (AttributeUtils::get(MPlug(globals, m_sppm_direct_lighting_mode), sppmDirectLightingMode))
+    {
+        if (sppmDirectLightingMode == 0)
```
Comment (dictoon) [inline src/appleseedmaya/renderglobalsnode.cpp:784]: Use a `switch`.

File: src/appleseedmaya/renderglobalsnode.cpp [hunk 12]
```diff
@@ -648,6 +767,97 @@ void RenderGlobalsNode::applyGlobalsToProject(
...
+    bool SPPMEnableIBL;
+    if (AttributeUtils::get(MPlug(globals, m_sppm_enable_ibl), SPPMEnableIBL))
+        INSERT_PATH_IN_CONFIGS("sppm.enable_ibl", SPPMEnableIBL)
+
+    bool limitPhotonTracingBounces = false;
+    AttributeUtils::get(MPlug(globals, m_sppm_photon_tracing_enable_bounce_limit), limitPhotonTracingBounces);
+
+    if (limitPhotonTracingBounces)
+    {
+        int phot_bounces = -1;
```
Comment (dictoon) [inline src/appleseedmaya/renderglobalsnode.cpp:807]: Avoid abbreviations whenever possible: `photon_bounces` (same below).

File: src/appleseedmaya/renderglobalsnode.cpp [hunk 13]
```diff
@@ -792,6 +1002,7 @@ void RenderGlobalsNode::applyGlobalsToProject(
 #undef REMOVE_PATH_IN_CONFIGS
 }
 
+// Post processing.
```
Comment (dictoon) [inline src/appleseedmaya/renderglobalsnode.cpp:1005]: *Post-processing*

**Review by dictoon — COMMENTED** [review]
Thanks!

File: scripts/appleseedMaya/renderGlobals.py (hunk 2 above)
Comment (Mango-3) [reply scripts/appleseedMaya/renderGlobals.py:502]: Done

File: scripts/appleseedMaya/renderGlobals.py (hunk 3 above)
Comment (Mango-3) [reply scripts/appleseedMaya/renderGlobals.py:517]: Unfortunately there is not enough space, so I have to abbreviate ussian Roulette Start Bounce with RR Start Bounce.

File: scripts/appleseedMaya/renderGlobals.py (hunk 4 above)
Comment (Mango-3) [reply scripts/appleseedMaya/renderGlobals.py:506]: Same as above

File: scripts/appleseedMaya/renderGlobals.py (hunk 5 above)
Comment (Mango-3) [reply scripts/appleseedMaya/renderGlobals.py:529]: Dito

File: src/appleseedmaya/renderglobalsnode.cpp (hunk 9 above)
Comment (Mango-3) [reply src/appleseedmaya/renderglobalsnode.cpp:718]: Replaced all multiple if with switch statements.

File: src/appleseedmaya/renderglobalsnode.cpp (hunk 12 above)
Comment (Mango-3) [reply src/appleseedmaya/renderglobalsnode.cpp:807]: Done

File: src/appleseedmaya/renderglobalsnode.cpp (hunk 13 above)
Comment (Mango-3) [reply src/appleseedmaya/renderglobalsnode.cpp:1005]: Good catch

**Comment by Mango-3** [comment]
The changes are in commit 4 of #189

**Pull request merged**